addon/adapters/rest.js

```diff
@@ -27,6 +27,26 @@ const {
   This adapter is designed around the idea that the JSON exchanged with
   the server should be conventional.
 
+  ## Success and failure
+
+  The REST adapter will consider a success any response with a status code
+  of the 2xx family ("Success"), as well as 304 ("Not Modified"). Any other
+  status code will be considered a failure.
+
+  On success, the request promise will be resolved with the full response
+  payload.
+
+  Failed responses with status code 422 ("Unprocessable Entity") will be
+  considered "invalid". The response will be discarded, except for the
+  `errors` key. The request promise will be rejected with a `DS.InvalidError`.
+  This error object will encapsulate the saved `errors` value.
+
+  Any other status codes will be treated as an "adapter error". The request
+  promise will be rejected, similarly to the "invalid" case, but with
+  an instance of `DS.AdapterError` instead. The rejection will receive
+  a second argument: an object with additional details about the
+  circumstances of the error.
```

Comment:
I think `The rejection will receive a second argument` is a little confusing here, as it sounds to me that the rejection handler takes a second argument: `.then(null, function(error, additonalDetails) {})`.

I think you meant that the `DS.AdapterError` is instantiated with a `message` as a second argument here, right? Or am I misunderstanding something?. I am looking [at this](https://github.com/emberjs/data/blob/e88dfed1e26b554c41029a218ea6dc41b7a786a6/addon/-private/adapters/errors.js#L14)...